File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/K._Ближайшая_остановка/solution.py
import sys
from collections import defaultdict

def euclidean_distance_squared(p1, p2):
    return (p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2

def solution(n_points, m_points):
    """
    Time Complexity O(N + M + NK)
    Memory Complexity O(N + M)
    """
    max_distance: int = 20
    grid_size: int = max_distance
    grid: dict[tuple[int, int], list[tuple[int, int]]] = defaultdict(list)

    for idx, (x, y) in enumerate(m_points):
        cell_x: int = x // grid_size
        cell_y: int = y // grid_size
        grid[(cell_x, cell_y)].append((x, y))

    max_count: int = 0
    best_point_index: int = -1

    for idx, (x_n, y_n) in enumerate(n_points):
        count: int = 0
        cell_x_n: int = x_n // grid_size
        cell_y_n: int = y_n // grid_size

        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                current_cell: tuple[int, int] = (cell_x_n + dx, cell_y_n + dy)
                if current_cell in grid:
                    for (x_m, y_m) in grid[current_cell]:
                        if euclidean_distance_squared((x_n, y_n), (x_m, y_m)) <= max_distance ** 2:
                            count += 1
        if count > max_count:
            max_count = count
            best_point_index = idx

    sys.stdout.write(f'{best_point_index + 1}\n')

if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    n_points: list[tuple[int, int]] = []
    for _ in range(n):
        x, y = map(int, sys.stdin.readline().rstrip().split())
        n_points.append((x, y))
    m: int = int(sys.stdin.readline().rstrip())
    m_points: list[tuple[int, int]] = []
    for _ in range(m):
        x, y = map(int, sys.stdin.readline().rstrip().split())
        m_points.append((x, y))
    solution(n_points, m_points)

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/K._Ближайшая_остановка/solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <tuple>
#include <cmath>
#include <functional>

struct TupleHash {
    template <class T>
    size_t operator()(const std::tuple<T, T>& t) const {
        auto hash1 = std::hash<T>()(std::get<0>(t));
        auto hash2 = std::hash<T>()(std::get<1>(t));
        return hash1 ^ (hash2 << 1);
    }
};

int euqlidieanDistanceSquared(std::tuple<int, int>& p1, std::tuple<int, int>& p2) {
    return std::pow(std::get<0>(p1) - std::get<0>(p2), 2) + std::pow(std::get<1>(p1) - std::get<1>(p2), 2);
}

void solution() {
    /*
    Time Complexity O(N + M + NK)
    Memory Complexity O(N + M)
    */
    int n;
    std::cin >> n;
    std::vector<std::tuple<int, int>> nPoints;
    for (int i = 0; i < n; ++i) {
        int x, y;
        std::cin >> x >> y;
        nPoints.push_back({x, y});
    }
    int m;
    std::cin >> m;
    std::vector<std::tuple<int, int>> mPoints;
    for (int i = 0; i < m; ++i) {
        int x, y;
        std::cin >> x >> y;
        mPoints.push_back({x, y});
    }
    int maxDistance = 20;
    int bound = std::pow(maxDistance, 2);
    int gridSize = maxDistance;
    std::unordered_map<std::tuple<int, int>, std::vector<std::tuple<int, int>>, TupleHash> grid;
    for (int idx = 0; idx < m; ++idx) {
        int cellX = std::get<0>(mPoints[idx]) / gridSize;
        int cellY = std::get<1>(mPoints[idx]) / gridSize;
        grid[{cellX, cellY}].push_back({std::get<0>(mPoints[idx]), std::get<1>(mPoints[idx])});
    }
    int maxCount = 0;
    int bestPointIndex = -1;
    std::vector<int> moves = {-1, 0, 1};
    for (int idx = 0; idx < n; ++idx) {
        int count = 0;
        int cellXN = std::get<0>(nPoints[idx]) / gridSize;
        int cellYN = std::get<1>(nPoints[idx]) / gridSize;
        for (int dX : moves) {
            for (int dY : moves) {
                std::tuple<int, int> currentCell = {cellXN + dX, cellYN + dY};
                if (grid.count(currentCell)) {
                    for (std::tuple<int, int> pair : grid[currentCell]) {
                        if (euqlidieanDistanceSquared(nPoints[idx], pair) <= bound) {
                            ++count;
                        }
                    }
                }
            }
        }
        if (count > maxCount) {
            maxCount = count;
            bestPointIndex = idx;
        }
    }
    std::cout << bestPointIndex + 1 << "\n";
}


int main() {
    solution();
}